File: scripts/statusline.js
#!/usr/bin/env node
/**
 * Node.js status line script for Claude Code
 * Usage: Copy to ~/.claude/statusline.js and make executable
 */

const { execSync } = require('child_process');
const path = require('path');

// ANSI Colors
const BLUE = '\x1b[0;34m';
const MAGENTA = '\x1b[0;35m';
const CYAN = '\x1b[0;36m';
const GREEN = '\x1b[0;32m';
const YELLOW = '\x1b[0;33m';
const RED = '\x1b[0;31m';
const DIM = '\x1b[2m';
const RESET = '\x1b[0m';

function getGitInfo(directory) {
    try {
        // Check if in git repo
        execSync('git rev-parse --git-dir', {
            cwd: directory,
            stdio: ['pipe', 'pipe', 'pipe']
        });

        // Get branch
        const branch = execSync('git branch --show-current', {
            cwd: directory,
            encoding: 'utf8'
        }).trim();

        if (!branch) return '';

        // Count changes
        const status = execSync('git status --porcelain', {
            cwd: directory,
            encoding: 'utf8'
        });
        const changes = status.split('\n').filter(l => l.trim()).length;

        if (changes > 0) {
            return ` | ${MAGENTA}${branch}${RESET} ${CYAN}[${changes}]${RESET}`;
        }
        return ` | ${MAGENTA}${branch}${RESET}`;
    } catch {
        return '';
    }
}

function formatTokens(tokens) {
    if (tokens >= 1000) {
        return `${(tokens / 1000).toFixed(1)}k`;
    }
    return String(tokens);
}

let input = '';

process.stdin.setEncoding('utf8');
process.stdin.on('data', chunk => input += chunk);

process.stdin.on('end', () => {
    let data;
    try {
        data = JSON.parse(input);
    } catch {
        console.log('[Claude] ~');
        return;
    }

    // Extract data
    const model = data.model?.display_name || 'Claude';
    const currentDir = data.workspace?.current_dir || '~';
    const dirName = path.basename(currentDir) || '~';

    // Context window
    const ctx = data.context_window || {};
    const contextSize = ctx.context_window_size || 200000;
    const inputTokens = ctx.total_input_tokens || 0;
    const outputTokens = ctx.total_output_tokens || 0;

    // Calculate used tokens
    const usedTokens = inputTokens + outputTokens;

    // Free tokens (matches /context "Free space" calculation)
    let freeTokens = contextSize - usedTokens;
    if (freeTokens < 0) {
        freeTokens = 0;
    }

    // Calculate percentage (matches /context output precision)
    const freePercent = contextSize > 0 ? (freeTokens * 100.0 / contextSize) : 0;

    // Autocompact is always enabled in Claude Code
    const autocompactEnabled = true;

    // Color based on free percentage
    let ctxColor;
    if (freePercent > 50) {
        ctxColor = GREEN;
    } else if (freePercent > 25) {
        ctxColor = YELLOW;
    } else {
        ctxColor = RED;
    }

    // Git info
    const gitInfo = getGitInfo(currentDir);

    // Cost
    const cost = data.cost?.total_cost_usd || 0;
    const costInfo = cost ? ` | ${DIM}$${cost.toFixed(4)}${RESET}` : '';

    // Autocompact indicator
    const autocompactInfo = autocompactEnabled ? `${DIM}[AC]${RESET}` : `${DIM}[AC:off]${RESET}`;

    // Output
    const freeDisplay = formatTokens(freeTokens);
    console.log(`${DIM}[${model}]${RESET} ${BLUE}${dirName}${RESET}${gitInfo} | ${ctxColor}${freeDisplay} free (${freePercent.toFixed(1)}%)${RESET} ${autocompactInfo}${costInfo}`);
});

File: scripts/statusline.py
#!/usr/bin/env python3
"""
Python status line script for Claude Code
Usage: Copy to ~/.claude/statusline.py and make executable
"""

import json
import sys
import os
import subprocess

# ANSI Colors
BLUE = '\033[0;34m'
MAGENTA = '\033[0;35m'
CYAN = '\033[0;36m'
GREEN = '\033[0;32m'
YELLOW = '\033[0;33m'
RED = '\033[0;31m'
DIM = '\033[2m'
RESET = '\033[0m'


def get_git_info(directory):
    """Get git branch and change count"""
    try:
        # Check if in git repo
        subprocess.run(
            ['git', '-C', directory, 'rev-parse', '--git-dir'],
            capture_output=True, check=True
        )

        # Get branch name
        result = subprocess.run(
            ['git', '-C', directory, 'branch', '--show-current'],
            capture_output=True, text=True
        )
        branch = result.stdout.strip()

        if not branch:
            return ""

        # Count changes
        result = subprocess.run(
            ['git', '-C', directory, 'status', '--porcelain'],
            capture_output=True, text=True
        )
        changes = len([l for l in result.stdout.split('\n') if l.strip()])

        if changes > 0:
            return f" | {MAGENTA}{branch}{RESET} {CYAN}[{changes}]{RESET}"
        return f" | {MAGENTA}{branch}{RESET}"
    except Exception:
        return ""


def format_tokens(tokens):
    """Format token count (k for thousands, with one decimal)"""
    if tokens >= 1000:
        return f"{tokens / 1000:.1f}k"
    return str(tokens)


def main():
    try:
        data = json.load(sys.stdin)
    except json.JSONDecodeError:
        print("[Claude] ~")
        return

    # Extract data
    model = data.get('model', {}).get('display_name', 'Claude')
    current_dir = data.get('workspace', {}).get('current_dir', '~')
    dir_name = os.path.basename(current_dir) or '~'

    # Context window
    ctx = data.get('context_window', {})
    context_size = ctx.get('context_window_size', 200000)
    input_tokens = ctx.get('total_input_tokens', 0)
    output_tokens = ctx.get('total_output_tokens', 0)

    # Calculate used tokens
    used_tokens = input_tokens + output_tokens

    # Free tokens (matches /context "Free space" calculation)
    free_tokens = context_size - used_tokens
    if free_tokens < 0:
        free_tokens = 0

    # Calculate percentage (matches /context output precision)
    free_percent = (free_tokens * 100.0 / context_size) if context_size > 0 else 0

    # Autocompact is always enabled in Claude Code
    autocompact_enabled = True

    # Color based on free percentage
    if free_percent > 50:
        ctx_color = GREEN
    elif free_percent > 25:
        ctx_color = YELLOW
    else:
        ctx_color = RED

    # Git info
    git_info = get_git_info(current_dir)

    # Cost
    cost = data.get('cost', {}).get('total_cost_usd', 0)
    cost_info = f" | {DIM}${cost:.4f}{RESET}" if cost else ""

    # Autocompact indicator
    autocompact_info = f"{DIM}[AC]{RESET}" if autocompact_enabled else f"{DIM}[AC:off]{RESET}"

    # Output
    free_display = format_tokens(free_tokens)
    print(f"{DIM}[{model}]{RESET} {BLUE}{dir_name}{RESET}{git_info} | {ctx_color}{free_display} free ({free_percent:.1f}%){RESET} {autocompact_info}{cost_info}")


if __name__ == '__main__':
    main()
